File: Trabajo_8/CPP/Actividades/Actividad_3.cpp
#include <iostream>

using namespace std;

class Animal {
public:
    virtual void hacerSonido() const = 0;
    virtual void moverse() const = 0;
    virtual ~Animal() = default;
};

class Perro : public Animal {
public:
    void hacerSonido() const override {
        cout << "El perro ladra: Guau guau!" << endl;
    }
    void moverse() const override {
        cout << "El perro corre." << endl;
    }
};

class Pajaro : public Animal {
public:
    void hacerSonido() const override {
        cout << "El pajaro canta: Pio pio!" << endl;
    }
    void moverse() const override {
        cout << "El pajaro vuela." << endl;
    }
};

int main() {
    Animal* ptrAnimal1;
    Animal* ptrAnimal2;

    Perro perro1;
    Pajaro pajaro1;

    ptrAnimal1 = &perro1;
    ptrAnimal2 = &pajaro1;

    ptrAnimal1->hacerSonido();
    ptrAnimal1->moverse();

    ptrAnimal2->hacerSonido();
    ptrAnimal2->moverse();

    return 0;
}

File: Trabajo_8/JAVA/Actividades/Actividad_3.java
package Trabajo_7.Actividades;

import java.util.ArrayList;
import java.util.List;


interface Animal {
    void hacerSonido();
    void moverse();
}

// Perro.java
class Perro implements Animal {
    @Override
    public void hacerSonido() {
        System.out.println("El perro ladra: ¡Guau guau!");
    }

    @Override
    public void moverse() {
        System.out.println("El perro corre.");
    }
}

class Pajaro implements Animal {
    @Override
    public void hacerSonido() {
        System.out.println("El pájaro canta: ¡Pío pío!");
    }

    @Override
    public void moverse() {
        System.out.println("El pájaro vuela.");
    }
}

public class Actividad_3 {
    public static void main(String[] args) {
        List<Animal> animales = new ArrayList<>();
        animales.add(new Perro());
        animales.add(new Pajaro());

        for (Animal animal : animales) {
            animal.hacerSonido();
            animal.moverse();
        }
    }
}
